data-processing/common/expand_macros.py

```diff
@@ -0,0 +1,492 @@
+                path = match.group("path").decode("utf-8", errors="ignore")
+                if path in used_in:
+                    end_line = int(match.group("end_line"))
+                    end_col = int(match.group("end_column"))
+
+        # If LaTeXML has finished expanding a control sequence, then mark the control sequence as
+        # having been expanded and pop it from the stack of macros to expand.
+        if event_type == "end-expansion":
+            cs_id = match.group("object_id")
+
+            cs = active_macros.get(cs_id)
+            if cs:
+                cs.expanded = True
+
+            if expanding == cs_id:
+                expanding = None
+
+
+def expand_macros(
+    contents: bytes,
+    expansions: List[Expansion],
+    wrap_expansions_in_groups: bool = False,
+) -> bytes:
+    """
+    Apply expansions to the contents of a TeX file. It is recommended that this method
+    is called with 'wrap_expansions_in_groups' set. This will wrap all expansions in
+    curly braces (i.e., '{' and '}', or group delimiters in TeX). The advantage of wrapping
+    in groups is that it averts several potential sources of TeX compilation errors that
+    could arise from expanding macros, including:
+
+    - expanding a macro right after another unexpanded macro such that it then gets
+      merged into the macro name before it
+      Example: \\unexpandable\\expandstox -> \\unexpandablex
+    """
+
+    # Sort expansions from last to first.
+    expansions_sorted = sorted(
+        expansions, key=lambda e: (e.start_line, e.start_col), reverse=True
+    )
+
+    # Map from expansions (by their index in the list of sorted expansions) to their
+    # start offsets and end offsets in the file (in number of bytes from the very
+    # start of the file; first character is at offset 0).
+    start_offsets: List[int] = [-1] * len(expansions_sorted)
+    end_offsets: List[int] = [-1] * len(expansions_sorted)
+
+    # Need to keep the line ends (i.e., newline characters at the ends of lines) as
+    # they should be included in the character offset.
+    lines = contents.splitlines(keepends=True)
+    line_start = 0
+    for line_number, line in enumerate(lines, start=1):
+        for ei, expansion in enumerate(expansions_sorted):
+            if start_offsets[ei] == -1 and expansion.start_line == line_number:
+                start_offsets[ei] = line_start + expansion.start_col
+            if end_offsets[ei] == -1 and expansion.end_line == line_number:
+                end_offsets[ei] = line_start + expansion.end_col
+
+        line_start += len(line)
+
+    # Apply expansions (in reverse order)
+    contents_copy = bytes(contents)
+    for ei, expansion in enumerate(expansions_sorted):
+        macro_start = start_offsets[ei]
+        macro_end = end_offsets[ei]
+        if macro_start == -1 or macro_end == -1:
+            continue
+        # Do not replace macro if the text to be expanded does not start with the macro name.
```

Comment:
would this indicate an error we should warn about or can come up normally?

Reply:
You are right, this would be a good thing to warn about (I would be interested in knowing if this issue ever occurs---right now I am only speculating that it could). I will add a warning 👍